File: android-client/app/src/main/java/com/cablecanvas/client/protocol/FrameProtocol.kt
package com.cablecanvas.client.protocol

import java.io.DataInputStream
import java.io.EOFException
import java.io.InputStream

object FrameProtocol {
    enum class Codec(val id: Int) {
        JPEG(1),
        H264(2),
    }

    data class Frame(
        val codec: Codec,
        val flags: Int,
        val width: Int,
        val height: Int,
        val payload: ByteArray,
    )

    private const val MAGIC = "CCF2"
    private const val MAX_FRAME_BYTES = 25 * 1024 * 1024

    fun readFrame(input: InputStream): Frame {
        val dataInput = DataInputStream(input)
        val magicBytes = ByteArray(4)
        dataInput.readFully(magicBytes)
        val actualMagic = String(magicBytes, Charsets.US_ASCII)
        if (actualMagic != MAGIC) {
            throw EOFException("Invalid frame magic: $actualMagic")
        }
        val codecRaw = dataInput.readUnsignedByte()
        val codec = Codec.entries.firstOrNull { it.id == codecRaw }
            ?: throw EOFException("Unknown codec id: $codecRaw")
        val flags = dataInput.readUnsignedByte()
        dataInput.readUnsignedShort() // reserved
        val width = dataInput.readInt()
        val height = dataInput.readInt()
        val length = dataInput.readInt()
        require(length in 1..MAX_FRAME_BYTES) { "Invalid frame length: $length" }
        val frame = ByteArray(length)
        dataInput.readFully(frame)
        return Frame(
            codec = codec,
            flags = flags,
            width = width,
            height = height,
            payload = frame,
        )
    }
}

File: mac-host/Sources/FrameProtocol.swift
import Foundation

enum FrameCodec: UInt8 {
    case jpeg = 1
    case h264 = 2
}

struct EncodedFrame {
    let codec: FrameCodec
    let flags: UInt8
    let width: Int
    let height: Int
    let payload: Data
}

enum FrameProtocol {
    private static let magic = Data("CCF2".utf8)

    static func encode(frame: EncodedFrame) -> Data {
        var header = magic
        header.append(frame.codec.rawValue)
        header.append(frame.flags)
        var reserved: UInt16 = 0
        withUnsafeBytes(of: &reserved) { bytes in
            header.append(contentsOf: bytes)
        }
        var width = UInt32(max(frame.width, 0)).bigEndian
        withUnsafeBytes(of: &width) { bytes in
            header.append(contentsOf: bytes)
        }
        var height = UInt32(max(frame.height, 0)).bigEndian
        withUnsafeBytes(of: &height) { bytes in
            header.append(contentsOf: bytes)
        }
        var length = UInt32(frame.payload.count).bigEndian
        withUnsafeBytes(of: &length) { bytes in
            header.append(contentsOf: bytes)
        }
        header.append(frame.payload)
        return header
    }
}
